### PR TITLE
Improve handling of failed JFR TaskRuns

### PR DESCRIPTION
**Pull request by anbrsap**
### Description

- Recreate JFR TaskRun if pod creation failed

    The creation of the JFR pod may temporarily fail, e.g. due to a
    timeout calling a mandatory admission webhook. Steward should detect
    this and recreate the Tekton taskrun to retry.

- Stop waiting for finished non-restartable JFR TaskRun

    If a JFR TaskRun was never started, is finished and is not
    restartable, Steward now fails the PipelineRun instead of waiting
    until timeout.

### Submitter checklist

- [x] Change has been tested (on a back-end cluster)
- [x] [changelog.yaml] entry with upgrade notes is prepared and appropriate for the audience affected by the change (users or developer, depending on the change)
- [x] Semantic version diffed against [last release][releases] and updated accordingly. In this project the version has to be maintained here:
    - [/charts/steward/Chart.yaml](https://github.com/SAP/stewardci-core/blob/master/charts/steward/Chart.yaml) (`version` and `appVersion`)

In case dependencies have been updated:
- [ ] Links to external changelogs, since the last release of our component, added to the [changelog.yaml] entry (description).
- [ ] Changelogs read thoroughly, potential impact described, upgrade notes prepared (if necessary)
- [ ] Check if dependency updates affect our semantic version increment.

### Reviewer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There is at least 1 approval for the pull request and no outstanding requests for change
- [x] All voter checks have passed
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] The Pull Request title is understandable and reflects the changes well
- [x] The Pull Request description is understandable and well documented
- [x] [changelog.yaml] entry for this Pull Request has been added
    - [x] Changelog entry contains all required information
    - [ ] 'Upgrade notes' are documented in changelog.yaml (if required)

[changelog.yaml]: https://github.com/SAP/stewardci-core/changelog.yaml
[releases]: https://github.com/SAP/stewardci-core/releases

### CODE REVIEW

File: pkg/runctl/controller.go [hunk 1]
```diff
@@ -585,10 +586,15 @@ func (c *Controller) handlePipelineRunWaiting(
...
-		if startTime.Add(timeout.Duration).Before(time.Now()) {
+
+		isWaitingTimeout := func() bool {
+			waitingStartTime := pipelineRun.GetStatus().StateDetails.StartedAt
+			timeout := c.getWaitTimeout(pipelineRunsConfig)
+			return waitingStartTime.Add(timeout.Duration).Before(time.Now())
+		}
+
+		failOnWaitingTimeout := func() (bool, error) {
+			timeout := c.getWaitTimeout(pipelineRunsConfig)
```
Comment (rinckm) [inline pkg/runctl/controller.go:597]:
```suggestion
	        timeout := c.getWaitTimeout(pipelineRunsConfig)
			
		isWaitingTimeout := func() bool {
			waitingStartTime := pipelineRun.GetStatus().StateDetails.StartedAt
			return waitingStartTime.Add(timeout.Duration).Before(time.Now())
		}

		failOnWaitingTimeout := func() (bool, error) {
```

File: pkg/runctl/run/interfaces.go [hunk 2]
```diff
@@ -21,7 +21,8 @@ type Manager interface {
 
 // Run represents a pipeline run
 type Run interface {
-	GetStartTime() *metav1.Time
+	IsDeleted() bool
+	GetJFRStepStartTime() *metav1.Time
```
Comment (rinckm) [inline pkg/runctl/run/interfaces.go:25]: Interface should not make assumptions about single steps in the run.

File: pkg/runctl/controller.go (hunk 1 above)
Comment (anbrsap) [reply pkg/runctl/controller.go]: Done

File: pkg/runctl/run/interfaces.go (hunk 2 above)
Comment (anbrsap) [reply pkg/runctl/run/interfaces.go]: I reverted the name change. I also moved the documentation to the interfaces.

File: pkg/runctl/controller.go [hunk 3]
```diff
@@ -585,42 +586,68 @@ func (c *Controller) handlePipelineRunWaiting(
...
 
-		started := run.GetStartTime()
-		if started != nil {
-			if err := c.changeAndCommitStateAndMeter(ctx, pipelineRun, api.StateRunning, *started); err != nil {
-				return true, err
+		taskRunFinished, _ := run.IsFinished()
+		if taskRunFinished /* without having started */ {
+			if run.IsRestartable() {
+				c.eventRecorder.Event(pipelineRun.GetReference(), corev1.EventTypeWarning, api.EventReasonWaitingFailed, "restarting")
+				return c.restartJFRTaskRun(ctx, runManager, pipelineRun)
```
Comment (rinckm) [inline pkg/runctl/controller.go:633]:
```suggestion
				return c.restartPipelineRun(ctx, runManager, pipelineRun)
```

**Review by rinckm — APPROVED** [review]

**Pull request merged**